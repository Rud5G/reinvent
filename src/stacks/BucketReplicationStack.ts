import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { S3 } from 'cdk-iam-floyd';
import { Construct } from 'constructs';

export interface BucketReplicationStackProps extends StackProps {
  buckets: string[];
}

export class BucketReplicationStack extends Stack {
  replicationRole: Role;

  constructor(scope: Construct, id: string, props: BucketReplicationStackProps) {
    super(scope, id, props);

    const replicationRole = this.replicationRole = new Role(this, 'ReplicationRole', {
      assumedBy: new ServicePrincipal('s3.amazonaws.com'),
    });

    let toList = new S3().toGetReplicationConfiguration().toListBucket();
    let toReplicate = new S3().toReplicateObject().toReplicateObject();
    let toGetObjectVersion = new S3().toGetObjectVersion().toGetObjectVersionAcl();

    for (const bucket of props.buckets) {
      toList.onBucket(bucket);
      toReplicate.onObject(bucket, 'converted/*');
      toGetObjectVersion.onObject(bucket, 'converted/*');
    }

    replicationRole.addToPolicy(toReplicate);
    replicationRole.addToPolicy(toList);
    replicationRole.addToPolicy(toGetObjectVersion);

    // we can't pass reference across stage boundaries, so this is going to get shared via a static string...
    new CfnOutput(this, 'ReplicationRoleArn', {
      value: replicationRole.roleArn,
    });
  }
}
