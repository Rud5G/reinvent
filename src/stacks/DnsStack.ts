import { Stack, StackProps } from 'aws-cdk-lib';
import { CfnRecordSet, CnameRecord, HostedZone } from 'aws-cdk-lib/aws-route53';
import { Construct } from 'constructs';

export interface DnsStackProps extends StackProps {
  websiteDomainName: string;
}

export class DnsStack extends Stack {
  constructor(scope: Construct, id: string, props: DnsStackProps) {
    super(scope, id, props);
    const hostedZone = HostedZone.fromLookup(this, 'HostedZone', { domainName: 'matthewbonig.com' });

    {
      const record = new CnameRecord(this, 'reinvent-us-east-1', {
        zone: hostedZone,
        recordName: `${props.websiteDomainName}`,
        domainName: `${props.websiteDomainName}-us-east-1.matthewbonig.com`,
      });
      const recordSet = (record.node.defaultChild as CfnRecordSet);
      recordSet.region = 'us-east-1';
      recordSet.setIdentifier = 'us-east-1';
    }
    {
      const record = new CnameRecord(this, 'reinvent-us-west-2', {
        zone: hostedZone,
        recordName: `${props.websiteDomainName}`,
        domainName: `${props.websiteDomainName}-us-west-2.matthewbonig.com`,
      });
      const recordSet = (record.node.defaultChild as CfnRecordSet);
      recordSet.region = 'us-west-2';
      recordSet.setIdentifier = 'us-west-2';
    }
  }


}
