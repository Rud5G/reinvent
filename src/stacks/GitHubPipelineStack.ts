import { App, Stack, StackProps, Tags } from 'aws-cdk-lib';
import { ContainerImage } from 'aws-cdk-lib/aws-ecs';
import { ShellStep } from 'aws-cdk-lib/pipelines';
import { AwsCredentials, GitHubWorkflow } from 'cdk-pipelines-github';
import { WebsiteProps } from './PipelineStack';
import { DEV_ACCOUNT, PRIMARY_REGION, PROD_ACCOUNT, SECONDARY_REGION } from '../constants';
import { ApplicationStage } from '../stages/ApplicationStage';
import { DnsStage } from '../stages/DnsStage';

interface GithubPipelineStackProps extends StackProps {
  replicationRoleArn: string;
  githubActionRoleArn: string;
  websiteImageAsset: ContainerImage;
}

export class GitHubPipelineStack extends Stack {
  constructor(scope: App, id: string, props: GithubPipelineStackProps) {
    super(scope, id, props);
    const { websiteImageAsset } = props;

    const devWebsiteProps: WebsiteProps = {
      imageAsset: websiteImageAsset,
      websiteDomainName: 'reinvent-dev',
      replicationRoleArn: props.replicationRoleArn,
    };
    const prodWebsiteProps: WebsiteProps = {
      imageAsset: websiteImageAsset,
      websiteDomainName: 'reinvent',
      replicationRoleArn: props.replicationRoleArn,
    };

    const pipeline = new GitHubWorkflow(scope, 'Pipeline', {
      awsCreds: AwsCredentials.fromOpenIdConnect({
        gitHubActionRoleArn: props.githubActionRoleArn,
      }),
      synth: new ShellStep('Synth', {
        commands: [
          'npm ci',
          'npm run build',
          'npx cdk synth',
        ],
      }),
    });

    const devWave = pipeline.addWave('Dev', {});

    const devPrimaryStage = new ApplicationStage(
      this,
      'DevPrimary',
      {
        env: {
          account: DEV_ACCOUNT,
          region: PRIMARY_REGION,
        },
        envTag: 'dev',
        ...devWebsiteProps,
        vpc: {
          cidr: '10.1.0.0/24',
        },
        isPrimary: true,
        tablePrefix: 'dev',
      });

    devWave.addStage(devPrimaryStage);

    const devSecondaryStage = new ApplicationStage(
      this,
      'DevSecondary',
      {
        env: {
          account: DEV_ACCOUNT,
          region: SECONDARY_REGION,
        },
        envTag: 'dev',
        ...devWebsiteProps,
        vpc: {
          cidr: '10.1.1.0/24',
        },
        isPrimary: false,
        tablePrefix: 'dev',
      });
    devWave.addStage(devSecondaryStage);

    pipeline.addStage(new DnsStage(
      this,
      'DevDns',
      {
        env: {
          account: DEV_ACCOUNT,
          region: PRIMARY_REGION,
        },
        websiteDomainName: devWebsiteProps.websiteDomainName,
      }),
    {
      post: [
        new ShellStep('Test Prod Site', {
          commands: [`curl https://${devWebsiteProps.websiteDomainName}.matthewbonig.com`],
        }),
      ],
    });

    const prodWave = pipeline.addWave('Prod', {});

    const prodPrimaryStage = new ApplicationStage(this, 'ProdPrimary', {
      env: {
        account: PROD_ACCOUNT,
        region: PRIMARY_REGION,
      },
      envTag: 'prod',
      ...prodWebsiteProps,
      vpc: {
        cidr: '10.2.0.0/24',
      },
      isPrimary: true,
      tablePrefix: 'prod',
    });

    prodWave.addStage(prodPrimaryStage, {});

    const prodSecondaryStage = new ApplicationStage(this, 'ProdSecondary', {
      env: {
        account: PROD_ACCOUNT,
        region: SECONDARY_REGION,
      },
      envTag: 'prod',
      ...prodWebsiteProps,
      vpc: {
        cidr: '10.2.1.0/24',
      },
      isPrimary: false,
      tablePrefix: 'prod',
    });

    prodWave.addStage(prodSecondaryStage, {});

    pipeline.addStage(new DnsStage(
      this,
      'ProdDns',
      {
        env: {
          account: PROD_ACCOUNT,
          region: PRIMARY_REGION,
        },
        websiteDomainName: prodWebsiteProps.websiteDomainName,
      }),
    {
      post: [
        new ShellStep('Test Dev Site', {
          commands: [`curl https://${prodWebsiteProps.websiteDomainName}.matthewbonig.com`],
        }),
      ],
    });

    pipeline.buildPipeline();

    Tags.of(prodSecondaryStage).add('test', 'test');
  }
}
