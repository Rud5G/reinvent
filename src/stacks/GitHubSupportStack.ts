import { GithubActionsIdentityProvider, GithubActionsRole } from 'aws-cdk-github-oidc';
import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { ManagedPolicy } from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

interface GithubSupportProps extends StackProps {
}

export class GitHubSupportStack extends Stack {
  constructor(scope: Construct, id: string, props: GithubSupportProps) {
    super(scope, id, props);

    const provider = GithubActionsIdentityProvider.fromAccount(
      this,
      'GithubProvider',
    );

    const deployRole = new GithubActionsRole(this, 'UploadRole', {
      provider: provider,
      owner: 'mbonig',
      repo: 'reinvent',
    });
    deployRole.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AdministratorAccess'));

    new CfnOutput(this, 'DeployRole', {
      value: deployRole.roleArn,
    });
  }
}
