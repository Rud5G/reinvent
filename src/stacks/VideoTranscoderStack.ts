import { Stack, StackProps, Tags } from 'aws-cdk-lib';
import { IBucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { Transcoder } from '../constructs/Transcoder';
import { TranscoderTests } from '../constructs/VideoTranscodingStackTest';

interface VideoTranscoderStackProps extends StackProps {
  mediaBucket: IBucket;
}

export class VideoTranscoderStack extends Stack {
  constructor(scope: Construct, id: string, props: VideoTranscoderStackProps) {
    super(scope, id, props);
    const { mediaBucket } = props;

    new Transcoder(this, 'VideoTranscoder', {
      bucket: mediaBucket,
    });

    new TranscoderTests(this, 'Tester', {
      bucket: mediaBucket,
    });

    Tags.of(this).add('role', 'video-transcoder');
  }
}


