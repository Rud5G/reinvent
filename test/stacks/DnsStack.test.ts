import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DnsStack } from '../../src/stacks/DnsStack';

test('DnsStack Snapshot', () => {
  const app = new App();
  const stack = new DnsStack(app, 'test', {
    websiteDomainName: 'reinvent.matthewbonig.com',
    env: {
      account: '000011112222',
      region: 'us-east-1',
    },

  });
  const assert = Template.fromStack(stack);

  expect(assert.toJSON()).toMatchSnapshot();
});
