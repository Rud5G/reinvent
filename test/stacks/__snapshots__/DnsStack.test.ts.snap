// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DnsStack Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "reinventuseast1EFD7B614": {
      "Properties": {
        "HostedZoneId": "DUMMY",
        "Name": "reinvent.matthewbonig.com.",
        "Region": "us-east-1",
        "ResourceRecords": [
          "reinvent.matthewbonig.com-us-east-1.matthewbonig.com",
        ],
        "SetIdentifier": "us-east-1",
        "TTL": "1800",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "reinventuswest2DA5BCC88": {
      "Properties": {
        "HostedZoneId": "DUMMY",
        "Name": "reinvent.matthewbonig.com.",
        "Region": "us-west-2",
        "ResourceRecords": [
          "reinvent.matthewbonig.com-us-west-2.matthewbonig.com",
        ],
        "SetIdentifier": "us-west-2",
        "TTL": "1800",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
